[gen_l10n] When localizing a message, prefer placeholder definitions defined by the current locale rather than the template locale

kzrnm opened the pull request:
- Fixes #153457
- Fixes #116716

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

kzrnm committed:
Add tests for fomatting DateTime with multiple locale

kzrnm committed:
Add a test for fomatting int with multiple locale

kzrnm committed:
[gen_l10n] Fix formatting with multiple locale

kzrnm committed:
Add tests

kzrnm committed:
Remove trailing space

andrewkolos commented on packages/flutter_tools/lib/src/localizations/gen_l10n_types.dart:
```diff
@@ -345,6 +345,7 @@ class Message {
       value = _value(templateBundle.resources, resourceId),
       description = _description(templateBundle.resources, resourceId, isResourceAttributeRequired),
       placeholders = _placeholders(templateBundle.resources, resourceId, isResourceAttributeRequired),
```
should we rename `placeholders` to `templatePlaceholders` to better differentiate between it and `localePlaceholders`?

kzrnm replied:
I renamed it.

andrewkolos commented:
Should https://github.com/flutter/flutter/blob/b364c07c1fa6241dfc4c7cb427837f54dff0e4b9/packages/flutter_tools/lib/src/localizations/gen_l10n_types.dart#L397 be updated to also consider placeholder definitions from non-template arb files?

kzrnm committed:
Merge branch 'master' into fixCustomDate

kzrnm committed:
Rename placeholders to templatePlaceholders

kzrnm committed:
Always import package:intl/intl.dart

kzrnm committed:
Generate parameters based on localePlaceholders

kzrnm commented:
@andrewkolos
I added support for cases where `format` is defined only in non-template and vice versa.

https://github.com/flutter/flutter/blob/d01e119f9cb95ec9f02319de1e8f933b0755edff/packages/flutter_tools/test/general.shard/generate_localizations_test.dart#L2683-L2759

There are two changes that accompany this change.

1. Always import package:intl/intl.dart
It is cumbersome to determine if it is necessary to import intl for each locale, so I changed it to always import it.
2. Compatibility Check
I added compatibility checks that will generate an error if a method is generated that is incompatible with the template. I am not a native English speaker, so if the error message is unnatural, please suggest a better fix.
https://github.com/flutter/flutter/blob/d01e119f9cb95ec9f02319de1e8f933b0755edff/packages/flutter_tools/lib/src/localizations/gen_l10n.dart#L131-L142

kzrnm committed:
Update packages/flutter_tools/lib/src/localizations/gen_l10n.dart

Co-authored-by: Andrew Kolos <andrewrkolos@gmail.com>

kzrnm commented on packages/flutter_tools/lib/src/localizations/gen_l10n.dart:
```diff
@@ -123,23 +123,36 @@ String _syntheticL10nPackagePath(FileSystem fileSystem) => fileSystem.path.join(
+  return message.templatePlaceholders.entries.map((MapEntry<String, Placeholder> e) {
+    final Placeholder placeholder = e.value;
+    final String? localePlaceholderType = localePlaceholders?[e.key]?.type;
+    if (localePlaceholderType != null && placeholder.type != localePlaceholderType) {
```
I would like to make the type explicit in this pull request. I think it would be better to loosen the type constraint in another pull request if necessary.

andrewkolos replied:
Right, i'm saying that I _want_ the type check to be strict, so I always want the type in the template ARB and the other ARBs to always match.

```suggestion
    if (placeholder.type != localePlaceholderType) {
```

kzrnm committed:
ignore: unused_import

kzrnm committed:
Update packages/flutter_tools/lib/src/localizations/gen_l10n.dart

Co-authored-by: Andrew Kolos <andrewrkolos@gmail.com>

andrewkolos committed:
Merge branch 'master' into pr/153459

bkonyi commented:
@kzrnm, FYI it looks like some checks are failing now.

kzrnm committed:
Merge remote-tracking branch 'upstream/master' into fixCustomDate

kzrnm committed:
Revert "Update packages/flutter_tools/lib/src/localizations/gen_l10n.dart"

This reverts commit e5c77427ef0f1c0e5d74288bbe9f5f9c88e7263b.

kzrnm committed:
Fix tests

kzrnm committed:
Add tests

kzrnm committed:
localePlaceholder.type should equal to templatePlaceholder.type

kzrnm commented on packages/flutter_tools/lib/src/localizations/gen_l10n_templates.dart:
```diff
@@ -171,7 +171,8 @@ const String dateVariableTemplate = '''
 
 const String classFileTemplate = '''
-@(header)@(requiresIntlImport)import '@(fileName)';
+@(header)import 'package:intl/intl.dart' as intl;
```
I understand. I have committed the fix.

kzrnm committed:
Merge branch 'master' into fixCustomDate

andrewkolos reviewed (approved):
LGTM. I will find a second reviewer soon.

(I also just restarted the Google testing check)

andrewkolos commented on packages/flutter_tools/lib/src/localizations/gen_l10n.dart:
```diff
@@ -123,23 +123,36 @@ String _syntheticL10nPackagePath(FileSystem fileSystem) => fileSystem.path.join(
+    final Placeholder placeholder = e.value;
+    final Placeholder? localePlaceholder = localePlaceholders?[e.key];
+    if (localePlaceholder != null && placeholder.type != localePlaceholder.type) {
+      throw L10nException(
```
Note for second reviewer:

This change _could_[^1] break peoples' builds. Consider

```json
// app_en.arb
{
  "@@locale": "en",
  "springBegins": "Spring begins on {springStartDate}",
  "@springBegins": {
    "description": "The first day of spring",
    "placeholders": {
      "springStartDate": {
        "type": "DateTime",
        "format": "MMMMd"
      }
    }
}
```

and
```json
// app_ja.arb
{
  "@@locale": "ja",
  "springBegins": "春が始まるのは{springStartDate}",
  "@springBegins": {
    "description": "The first day of spring",
    "placeholders": {
      "springStartDate": {
        "type": "int", // Type change.
        "format": "compactCurrency"
      }
    }
  },
```

Then, trying to regenerate localization code will result in a failure:

```
Generating synthetic localizations package failed with 1 error:

Error: The placeholder, springStartDate, has its "type" resource attribute set to the "int" type in locale "ja", but it is "DateTime" in the template placeholders. For
compatibility with template placeholders, change the "type" attribute to "DateTime".
```

I think this failing the build makes sense here, and I believe there's no pre-existing behavior for users to be dependent on here (the `springStartDate` placeholder in the `ja` locale would have been silently ignored, hence this PR). I'm content with this landing without a breaking change announcement.

[^1]: And therefore _will_ 🙃

bkonyi replied:
I think that's reasonable.

andrewkolos commented on packages/flutter_tools/lib/src/localizations/gen_l10n.dart:
```diff
@@ -123,23 +123,36 @@ String _syntheticL10nPackagePath(FileSystem fileSystem) => fileSystem.path.join(
+    final Placeholder? localePlaceholder = localePlaceholders?[e.key];
+    if (localePlaceholder != null && placeholder.type != localePlaceholder.type) {
+      throw L10nException(
+          'The placeholder, ${placeholder.name}, has its "type" resource attribute set to '
+          'the "${localePlaceholder.type}" type in locale "$locale", but it is "${placeholder.type}" '
+          'in the template placeholders. For compatibility with template placeholders, change '
+          'the "type" attribute to "${placeholder.type}".');
```
Nitpick: I think `placeholders` should be in its singular form here:

```suggestion
          'The placeholder, ${placeholder.name}, has its "type" resource attribute set to '
          'the "${localePlaceholder.type}" type in locale "$locale", but it is "${placeholder.type}" '
          'in the template placeholder. For compatibility with template placeholder, change '
          'the "type" attribute to "${placeholder.type}".');
```

kzrnm committed:
Update packages/flutter_tools/lib/src/localizations/gen_l10n.dart

Co-authored-by: Andrew Kolos <andrewrkolos@gmail.com>

bkonyi committed:
Update generate_localizations_test.dart

bkonyi reviewed (approved):
LGTM! Thanks for the contribution. I'll land this when the checks pass.

andrewkolos committed:
Merge branch 'master' into fixCustomDate
